File: 05-behavioral-patterns/02-observer/02-observer.go
package main

import "fmt"

type Observer interface {
	Update(news string)
}

type Subject interface {
	Attach(o Observer)
	Detach(o Observer)
	Notify(news string)
}

type NewsAgency struct {
	observers []Observer
}

func (n *NewsAgency) Attach(o Observer) {
	n.observers = append(n.observers, o)
}

func (n *NewsAgency) Detach(o Observer) {
	for i, observer := range n.observers {
		if observer == o {
			n.observers = append(n.observers[:i], n.observers[i+1:]...)
			break
		}
	}
}

func (n *NewsAgency) Notify(news string) {
	for _, observer := range n.observers {
		observer.Update(news)
	}
}

func (n *NewsAgency) PublishNews(news string) {
	fmt.Println("News published:", news)
	n.Notify(news)
}

type Subscriber struct {
	name string
}

func (s *Subscriber) Update(news string) {
	fmt.Printf("%s received news: %s\n", s.name, news)
}

func main() {
	agency := &NewsAgency{}
	alice := &Subscriber{name: "Alice"}
	bob := &Subscriber{name: "Bob"}

	agency.Attach(alice)
	agency.Attach(bob)

	agency.PublishNews("Observer Pattern Implemented!")
}

File: 05-behavioral-patterns/02-observer/02-observer.py
from typing import List

class Observer:
    def update(self, message: str):
        pass

class Subject:
    def __init__(self):
        self._observers: List[Observer] = []

    def attach(self, observer: Observer):
        self._observers.append(observer)

    def detach(self, observer: Observer):
        self._observers.remove(observer)

    def notify(self, message: str):
        for observer in self._observers:
            observer.update(message)

class NewsAgency(Subject):
    def publish_news(self, news: str):
        print(f"News published: {news}")
        self.notify(news)

class Subscriber(Observer):
    def __init__(self, name: str):
        self.name = name

    def update(self, message: str):
        print(f"{self.name} received news: {message}")

# Usage
agency = NewsAgency()
s1 = Subscriber("Alice")
s2 = Subscriber("Bob")

agency.attach(s1)
agency.attach(s2)

agency.publish_news("Observer Pattern Implemented!")
